proposals/0167-loader-v4.md

```diff
@@ -0,0 +1,343 @@
+- Instruction data:
+  - Enum variant `1u32`
+  - `u32` Byte offset at which to write
+  - `u32` Byte offset at which to read
+  - `u32` Length of the chunk to copy in bytes
+- Behavior:
+  - Check there are at least three instruction accounts,
+  otherwise throw `NotEnoughAccountKeys`
+  - Check that program account and source account do not alias,
+  otherwise throw `AccountBorrowFailed`
+  - Verify the program account
+  - Check the status stored in the program account is retracted,
+  otherwise throw `InvalidArgument`
+  - Check that the source account is owned by loader v1, v2, v3 or v4,
+  otherwise throw `InvalidArgument`
+  - and look-up the source header size:
+    - loader-v1: 0 bytes
+    - loader-v2: 0 bytes
+    - loader-v3: 45 bytes
+    - loader-v4: 48 bytes
+  - Check that the source end offset (sum of source offset and length) does
+  not exceed the maximum (source account length minus the source header size),
+  otherwise throw `AccountDataTooSmall`
+  - Check that the destination end offset (sum of destination offset and
+  length) does not exceed the maximum (program account length minus the loader-v4
+  header size), otherwise throw `AccountDataTooSmall`
+  - Copy the chunk between the program accounts at the offsets, each shifted by
+  the header size of their loader (account owner) respectively
+
+#### SetProgramLength
+
+- Instruction accounts:
+  - `[writable]` The program account to change the size of.
+  - `[signer]` The authority of the program.
+  - `[writable]` Optional, the recipient account.
+- Instruction data:
+  - Enum variant `2u32`
+  - `u32` The new size after the operation.
+- Behavior:
+  - Check there are at least two instruction accounts,
+    otherwise throw `NotEnoughAccountKeys`
+  - If this is an initialization (program account length is too short to
+  contain the header):
+    - the owner of the program account is loader-v4,
+    otherwise throw `InvalidAccountOwner`
+    - the program account is writable, otherwise throw `InvalidArgument`
+    - the provided authority (instruction account at index 1) signed,
+    otherwise throw `MissingRequiredSignature`
+  - If this is not an initialization:
+    - Verify the program account
+    - Check that the status stored in the program account is retracted,
+    otherwise throw `InvalidArgument`
+  - Check that there are enough funds in the program account for rent
+  exemption, otherwise throw `InsufficientFunds`
+  - If there are more than enough funds:
+    - Check there are at least three instruction accounts,
+    otherwise throw `NotEnoughAccountKeys`
+    - Check that the recipient account (instruction account at index 2) is
+    writable, otherwise throw `InvalidArgument`
+    - If a recipient account was provided that is not the program account:
+      - Transfer the surplus from the program account to the recipient account
+    - otherwise, if the requested new size is zero throw `InvalidArgument`
+  - If the requested new size is zero:
```

Comment:
Can you specify more explicitly how a program account should be deleted?
1. reduce lamports to zero
2. set data length to zero
3. possibly reset `is_executable` flag

Reply:
> The account is deleted implicitly at the end of the transaction 

It might not be implicitly deleted at the end of the transaction if its lamport balance is topped up by a later instruction in the transaction. The loader-v4 program cannot guarantee that the program account will be deleted. All it can guarantee is the state of certain account fields like the data, lamports, owner, and executable flag. So I think this SIMD should be explicit about what exactly the loader-v4 program will be changing about the account rather than stating a (potentially false) assumption about the account getting deleted later.

Reply:
That would be nice and also please list the account modifications to make the spec explicit (lamports, data length, etc)